### PR TITLE
All the changes committed.

### PR DESCRIPTION
**Pull request by sushantyelpale**

### CODE REVIEW

File: DALOptimizer/CSharpFile.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace DALOptimizer
+{
+    public class CSharpFile
+    {
+        public readonly CSharpProject Project;
+        public readonly string FileName;
+        public readonly string OriginalText;
+
+        public SyntaxTree SyntaxTree;
```
Comment (kapilbhosale) [inline DALOptimizer/CSharpFile.cs]: change it variable name must not be class name.

Comment (kapilbhosale) [reply DALOptimizer/CSharpFile.cs]: http://www.oracle.com/technetwork/java/codeconventions-135099.html

File: DALOptimizer/CSharpFile.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            this.FileName = fileName;
+
+            CSharpParser p = new CSharpParser(project.CompilerSettings);
+            //			using (var stream = File.OpenRead(fileName)) {
+            //				this.CompilationUnit = p.Parse(stream, fileName);
+            //			}
+
+            // Keep the original text around; we might use it for a refactoring later
+            this.OriginalText = File.ReadAllText(fileName);
+            this.SyntaxTree = p.Parse(this.OriginalText, fileName);
```
Comment (kapilbhosale) [inline DALOptimizer/CSharpFile.cs]: Method logic should not be constructor.

File: DALOptimizer/CSharpProject.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		/// </summary>
+		public readonly IProjectContent ProjectContent;
+		
+		/// <summary>
+		/// The resolved type system for this project.
+		/// This field is initialized once all projects have been loaded (in Solution constructor).
+		/// </summary>
+		public ICompilation Compilation;
+		
+		public CSharpProject(Solution solution, string title, string fileName)
```
Comment (kapilbhosale) [inline DALOptimizer/CSharpProject.cs:53]: Make Functions for the code, don't write it in a constructor.

File: DALOptimizer/MatchExpr.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
Comment (kapilbhosale) [inline DALOptimizer/MatchExpr.cs]: Remove unused files.

File: DALOptimizer/MatchInvocation.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+namespace DALOptimizer
+{
+    class MatchInvocation
+    {
+        AllPatterns Pat = new AllPatterns();
+        //PrintFunction PrFun = new PrintFunction();
+        // For All Global field Declarations
+        public void MatchFieldDecl(AstNode invocation, CSharpFile file, CSharpAstResolver astResolver)
+        {
+            if (invocation.Parent.GetType().Name.ToString() == "TypeDeclaration")
```
Comment (kapilbhosale) [inline DALOptimizer/MatchInvocation.cs]: Use string from constant variables.

File: DALOptimizer/MatchInvocation.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            var expr4 = Pat.StoredProc();
+            var expr5 = Pat.sqlConnstmt();
+            var expr6 = Pat.ConnOpenExprStmt();
+            var expr7 = Pat.ConnCloseExprStmt();
+            var expr8 = Pat.CmdDisposeExprStmt();
+            var expr9 = Pat.SqlDataAdapterExprStmt();
+            var expr10 = Pat.ExNonQuery1();
+            var expr11 = Pat.ConvertToInt32();
+
+            Match FillExpr = expr1.Match(invocation);
```
Comment (kapilbhosale) [inline DALOptimizer/MatchInvocation.cs]: avoid using unused var

File: DALOptimizer/MatchInvocation.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            Match logErr = expr2.Match(invocation);
```
Comment (kapilbhosale) [inline DALOptimizer/MatchInvocation.cs]: Avoid extra match if true is found.

File: DALOptimizer/Program.cs [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+using ICSharpCode.NRefactory.Editor;
+using ICSharpCode.NRefactory.PatternMatching;
+using ICSharpCode.NRefactory.TypeSystem;
+
+namespace DALOptimizer
+{
+    class Program
+    {
+        [STAThread]
+        public static void Main(string[] args)
```
Comment (kapilbhosale) [inline DALOptimizer/Program.cs:17]: Avoid writing logic in main.

File: DALOptimizer/Program.cs [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+            {
+
+/*                if (file.FileName != "D:\\DALProject\\DAL\\RecharegeReportDAL.cs")
+                {
+                    continue;
+                }
+                */
+
+                if (Path.GetFileName(file.FileName) == "DatabaseProcessing.cs")
+                {
```
Comment (kapilbhosale) [inline DALOptimizer/Program.cs]: dont use parenthesis if not needed.

File: DALOptimizer/Program.cs [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+                    //trak all expression Statemens
+                    if (invocation.GetType().Name == "ExpressionStatement")
+                    {
+                        Mr.MatchExprStmt(invocation, file, astResolver);
+                        continue;
+                    }
+                    //catch clause
+                    if (invocation.GetType().Name == "CatchClause")
```
Comment (kapilbhosale) [inline DALOptimizer/Program.cs]: var nodeTypes = ["CatchClause", "FieldDeclaration"];
foreach (var nodeType in nodeTypes){
if (invocation.GetType().Name == nodeType)
 +                    {
 +                        Mr.MatchExprStmt(invocation, file, astResolver);
 +                        continue;
 +                    }
}

File: DALOptimizer/Program.cs [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+            //pr.PrintMethod(solution);
+
+            Console.Write("Apply refactorings? ");
+            string answer = Console.ReadLine();
+            if ("yes".Equals(answer, StringComparison.OrdinalIgnoreCase) || "y".Equals(answer, StringComparison.OrdinalIgnoreCase))
+            {
+                foreach (var file in solution.AllFiles)
+                {
+
+                    if (file.IndexOfFieldDecl.Count == 0 && file.IndexOfPropDecl.Count == 0 &&
```
Comment (kapilbhosale) [inline DALOptimizer/Program.cs]: don't write very long expressions.

File: DALOptimizer/Program.cs [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+                                int start = script.GetCurrentOffset(Tempvar.PrevSibling.StartLocation);
+                                int end = script.GetCurrentOffset(Tempvar.EndLocation);
+                                script.RemoveText(start, end - start);
+                            }
+                            if (SqlDtAdptStmt.Success)
+                            {
+                                string MthdretType = expr.GetParent<MethodDeclaration>().ReturnType.GetText();
+                                if (MthdretType == "DataTable" || MthdretType == "System.Data.DataTable")
+                                {
+                                    string varName = expr.GetText().Split("()".ToCharArray())[1];
```
Comment (kapilbhosale) [inline DALOptimizer/Program.cs]: add classes and separate methods.

File: DALOptimizer/Program.cs [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+                            var expr5 = Pat.ExNonQuery();
+                            var expr6 = Pat.ConnOpenExprStmt();
+                            var expr7 = Pat.ConnCloseExprStmt();
+                            var expr8 = Pat.CmdDisposeExprStmt();
+                            var expr9 = Pat.SqlDataAdapterExprStmt();
+
+                            var expr12 = Pat.ExNonQuery1();
+                            var expr13 = Pat.ConvertToInt32();
+
+                            ICSharpCode.NRefactory.PatternMatching.Match FillExpr = expr2.Match(expr);
```
Comment (kapilbhosale) [inline DALOptimizer/Program.cs]: if (Pat.FillExpr().Match(expr).Success){
script.Remove(expr, true);
continue;
}

File: DALOptimizer/Solution.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+                Match match = projectLinePattern.Match(line);
+                if (match.Success)
+                {
+                    string typeGuid = match.Groups["TypeGuid"].Value;
+                    string title = match.Groups["Title"].Value;
+                    string location = match.Groups["Location"].Value;
+                    string guid = match.Groups["Guid"].Value;
+                    switch (typeGuid.ToUpperInvariant())
+                    {
+                        case "{2150E333-8FDC-42A3-9474-1A3956D46DE8}": // Solution Folder
```
Comment (kapilbhosale) [inline DALOptimizer/Solution.cs]: use `2150E333-8FDC-42A3-9474-1A3956D46DE81` as constant.

**Review by kapilbhosale — COMMENTED** [review]
@sushantyelpale Use OOPs properly.

File: DALOptimizer/DALOpt.csproj [hunk 15]
```diff
@@ -0,0 +1,130 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (akshaymohite) [inline DALOptimizer/DALOpt.csproj:1]: read about `gitignore`

File: DALOptimizer/MatchExpr.cs (hunk 4 above)
Comment (sushantyelpale) [reply DALOptimizer/MatchExpr.cs]: Removed All unused Files

File: DALOptimizer/MatchInvocation.cs (hunk 6 above)
Comment (sushantyelpale) [reply DALOptimizer/MatchInvocation.cs]: removed all unused Variables

File: DALOptimizer/Program.cs (hunk 8 above)
Comment (sushantyelpale) [reply DALOptimizer/Program.cs:17]: Removed all Logic from Main

File: DALOptimizer/CSharpFile.cs [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        public readonly CSharpProject Project;
+        public readonly string FileName;
+        public readonly string OriginalText;
+
+        public SyntaxTree syntaxTree;
+        public CSharpUnresolvedFile UnresolvedTypeSystemForFile;
+
+        public CSharpFile(CSharpProject project, string fileName)
+        {
+            this.Project = project;
```
Comment (akshaymohite) [inline DALOptimizer/CSharpFile.cs]: can be renamed as `this.project`

File: DALOptimizer/CSharpFile.cs [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+            this.FileName = fileName;
```
Comment (akshaymohite) [inline DALOptimizer/CSharpFile.cs]: can be renamed as `this.fileName`

File: DALOptimizer/CSharpFile.cs [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+            CSharpParser parsedFile = new CSharpParser(project.CompilerSettings);
+
+            // Keep the original text around; it might get used for refactoring later.
+            this.OriginalText = File.ReadAllText(fileName);
```
Comment (akshaymohite) [inline DALOptimizer/CSharpFile.cs]: can be renamed as `originalText`

File: DALOptimizer/CSharpFile.cs [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+            this.syntaxTree = parsedFile.Parse(this.OriginalText, fileName);
+
+            if (parsedFile.HasErrors)
+            {
+                Console.WriteLine("Error parsing " + fileName + ":");
+                foreach (var error in parsedFile.ErrorsAndWarnings)
+                    Console.WriteLine("  " + error.Region + " " + error.Message);
+            }
+            this.UnresolvedTypeSystemForFile = this.syntaxTree.ToTypeSystem();
```
Comment (akshaymohite) [inline DALOptimizer/CSharpFile.cs]: can be renamed as `unresolvedTypeSystemForFile`

File: DALOptimizer/CSharpProject.cs [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+		/// <summary>
+		/// The unresolved type system for this project.
+		/// </summary>
+		public readonly IProjectContent ProjectContent;
+		
+		/// <summary>
+		/// The resolved type system for this project.
+		/// This field is initialized once all projects have been loaded (in Solution constructor).
+		/// </summary>
+		public ICompilation Compilation;
```
Comment (akshaymohite) [inline DALOptimizer/CSharpProject.cs:51]: @sushantyelpale rename all variable names in camel case with first letter in lower case.

File: DALOptimizer/MatchInvocation.cs [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+            else if (Pat.ExNonQuery().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.StoredProc().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.sqlConnstmt().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.ConnOpenExprStmt().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.ConnCloseExprStmt().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.CmdDisposeExprStmt().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.SqlDataAdapterExprStmt().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.ExNonQuery1().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
+            else if (Pat.ConvertToInt32().Match(invocation).Success) { file.IndexOfExprStmt.Add((ExpressionStatement)invocation); }
```
Comment (akshaymohite) [inline DALOptimizer/MatchInvocation.cs]: @sushantyelpale wouldn't using `switch` be a better choice here?

File: DALOptimizer/MatchExpr.cs [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+                AssignmentExpr(script, file);
+                VarDeclarationStmt(script, file);
+                ExprStatement(script, file);
+                MethodDecl(script, file);
+            }
+            
+            return document;
+        }
+
+        private void GlobalFieldDeclaraton(DocumentScript script, CSharpFile file)
```
Comment (akshaymohite) [inline DALOptimizer/MatchExpr.cs]: type near `Declaraton`, probably `Declaration`?

File: DALOptimizer/MatchInvocation.cs [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+                }
+
+                //check Finally Block
+                if (invocation.GetType().Name == BLOCKSTATEMENT)
+                {
+                    MatchBlock(invocation, file, astResolver);
+                    continue;
+                }
+
+                if (invocation.GetType().Name == METHODDECLARARION)
```
Comment (akshaymohite) [inline DALOptimizer/MatchInvocation.cs]: @sushantyelpale no need to calculate `invocation.GetType().Nam` for all the `if` condition checks.
Calculate it once and then use `switch` case to perform the necessary operation.

File: DALOptimizer/MatchInvocation.cs [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+
+        public void MatchCatchClause(AstNode invocation, CSharpFile file, CSharpAstResolver astResolver)
+        {
+            file.IndexOfCtchClause.Add((CatchClause)invocation);
+        }
+
+        public void MatchAssExpr(AstNode invocation, CSharpFile file, CSharpAstResolver astResolver)
+        {
+            Match sqlCmdstmt1 = Pat.sqlCmdstmt1().Match(invocation);
+            if (sqlCmdstmt1.Success)
```
Comment (akshaymohite) [inline DALOptimizer/MatchInvocation.cs]: @sushantyelpale rename `sqlCmdstmt1` to make it more meaningful. Variables should be named which we can pronounce.

File: DALOptimizer/MatchInvocation.cs [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+                file.IndexOfAssExpr.Add((AssignmentExpression)invocation);
+        }
+
+        public void MatchVarDeclStmt(AstNode invocation, CSharpFile file, CSharpAstResolver astResolver)
+        {
+            Match sqlCmdstmt2 = Pat.sqlCmdstmt2().Match(invocation);
```
Comment (akshaymohite) [inline DALOptimizer/MatchInvocation.cs]: @sushantyelpale rename `sqlCmdstmt2` to make it more meaningful. Variables should be named which we can pronounce.

File: DALOptimizer/PrintFunction.cs [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace DALOptimizer
+{
+    class PrintFunction
+    {
+        public void PrintMethod(Solution solution)
+        {
+            Console.WriteLine("Found {0} places to refactor Field Declaration in {1} files.",
+                  solution.AllFiles.Sum(f => f.IndexOfFieldDecl.Count),
+                  solution.AllFiles.Count(f => f.IndexOfFieldDecl.Count > 0));
```
Comment (akshaymohite) [inline DALOptimizer/PrintFunction.cs:14]: @sushantyelpale `f.IndexOfFieldDecl.Count` is performed twice, try to perform it only once if it recalculates it every time we call it.

File: DALOptimizer/PrintFunction.cs [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+            Console.WriteLine("Found {0} places to refactor Expression statement in {1} files.",
+                  solution.AllFiles.Sum(f => f.IndexOfExprStmt.Count),
+                  solution.AllFiles.Count(f => f.IndexOfExprStmt.Count > 0));
+            Console.WriteLine("Found {0} places to refactor Catch clause in {1} files.",
+                  solution.AllFiles.Sum(f => f.IndexOfCtchClause.Count),
+                  solution.AllFiles.Count(f => f.IndexOfCtchClause.Count > 0));
+            Console.WriteLine("Found {0} places to refactor Method Declaration in {1} files.",
+                  solution.AllFiles.Sum(f => f.IndexOfMthdDecl.Count),
+                  solution.AllFiles.Count(f => f.IndexOfMthdDecl.Count > 0));
+            Console.WriteLine("Found {0} places to refactor Variable Declaration in {1} files.",
```
Comment (akshaymohite) [inline DALOptimizer/PrintFunction.cs:33]: @sushantyelpale try to see if all these `Console.WriteLine` could be done dynamically (probably calling in a loop)?

File: DALOptimizer/Program.cs [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+using ICSharpCode.NRefactory.TypeSystem;
+
+namespace DALOptimizer
+{
+    class Program
+    {
+        [STAThread]
+        public static void Main(string[] args)
+        {
+            OpenFileDialog opnfd = new OpenFileDialog();
```
Comment (akshaymohite) [inline DALOptimizer/Program.cs]: @sushantyelpale rename `opnfd` to make it more meaningful.

File: DALOptimizer/Program.cs [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+                if (Path.GetFileName(file.FileName) == "DatabaseProcessing.cs")
+                       continue;
+                var astResolver = new CSharpAstResolver(file.Project.Compilation, file.syntaxTree, file.UnresolvedTypeSystemForFile);
+                matchInvocation.CheckAllInvocation(file, astResolver);
+            }
+
+            //PrintFunction pr = new PrintFunction();
+            //pr.PrintMethod(solution);
+
+            Console.Write("Apply refactorings? ");
```
Comment (akshaymohite) [inline DALOptimizer/Program.cs]: @sushantyelpale probably we should suggest here as end user has to type "yes" or "y" to apply refctoring?

File: DALOptimizer/Solution.cs [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+        public readonly List<CSharpProject> Projects = new List<CSharpProject>();
+
+        public IEnumerable<CSharpFile> AllFiles
+        {
+            get{  return Projects.SelectMany(p => p.Files); }
+        }
+
+        public Solution(string fileName)
+        {
+            this.Directory = Path.GetDirectoryName(fileName);
```
Comment (akshaymohite) [inline DALOptimizer/Solution.cs]: `this.directory`

File: DALOptimizer/Solution.cs [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+                    Match match = projectLinePattern.Match(line);
+                    if (match.Success)
+                    {
+                        string typeGuid = match.Groups["TypeGuid"].Value;
+                        string title = match.Groups["Title"].Value;
+                        string location = match.Groups["Location"].Value;
+                        string guid = match.Groups["Guid"].Value;
+                        switch (typeGuid.ToUpperInvariant())
+                        {
+                            case "{2150E333-8FDC-42A3-9474-1A3956D46DE8}": // Solution Folder
```
Comment (akshaymohite) [inline DALOptimizer/Solution.cs]: @sushantyelpale why are these hard codings?

File: DALOptimizer/Solution.cs [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+                                // ignore folders
+                                break;
+                            case "{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}": // C# project
```
Comment (akshaymohite) [inline DALOptimizer/Solution.cs]: @sushantyelpale why are these hard codings?

File: DALOptimizer/bin/Debug/DALOpt.exe.config [hunk 33]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+	<startup>
+		<supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.0" />
+	</startup>
+</configuration>
```
Comment (akshaymohite) [inline DALOptimizer/bin/Debug/DALOpt.exe.config:6]: fix line endings in the editor.

File: DALOptimizer/MatchExpr.cs [hunk 34]
```diff
@@ -0,0 +1,265 @@
...
+                var copy = (ExpressionStatement)expr.Clone();
+
+                AllPatterns Pat = new AllPatterns();
+                if (Pat.FillExpr().Match(expr).Success) { script.Remove(expr, true); continue; }
+                if (Pat.StoredProc().Match(expr).Success) { script.Remove(expr, true); continue; }
+                if (Pat.sqlConnstmt().Match(expr).Success) { script.Remove(expr, true); continue; }
+                if (Pat.ConnOpenExprStmt().Match(expr).Success) { script.Remove(expr, true); continue; }
+                if (Pat.ConnCloseExprStmt().Match(expr).Success) { script.Remove(expr, true); continue; }
+                if (Pat.CmdDisposeExprStmt().Match(expr).Success) { script.Remove(expr, true); continue; }
+                if (Pat.ConvertToInt32().Match(expr).Success) { script.Remove(expr, true); continue; }
```
Comment (akshaymohite) [inline DALOptimizer/MatchExpr.cs]: The block `{ script.Remove(expr, true); continue; }` is same for all `if` conditions. Can be done in a single line.

File: DALOptimizer/MatchExpr.cs [hunk 35]
```diff
@@ -0,0 +1,265 @@
...
+        }
+        private void ExprStatement(DocumentScript script, CSharpFile file)
+        {
+            foreach (var expr in file.IndexOfExprStmt)
+            {
+                var copy = (ExpressionStatement)expr.Clone();
+
+                AllPatterns Pat = new AllPatterns();
+                if (Pat.FillExpr().Match(expr).Success) { script.Remove(expr, true); continue; }
+                if (Pat.StoredProc().Match(expr).Success) { script.Remove(expr, true); continue; }
```
Comment (akshaymohite) [inline DALOptimizer/MatchExpr.cs]: Let's say `Pat.FillExpr().Match(expr).Success` evaluates to `true` and `script.Remove(expr, true); continue;` is performed.

Then, do we need to perform all these other `if` conditions?

File: DALOptimizer/MatchExpr.cs [hunk 36]
```diff
@@ -0,0 +1,265 @@
...
+                    {
+                        string varName = expr.GetText().Split("()".ToCharArray())[1];
+                        var getDtTbl = Pat.GetDtTbl(varName);
+                        script.Replace(expr, getDtTbl);
+                    }
+                    else if (retType == "DataSet" || retType == "System.Data.DataSet")
+                    {
+                        string varName = expr.GetText().Split("()".ToCharArray())[1];
+                        var getDtSet = Pat.GetDtSet(varName);
+                        script.Replace(expr, getDtSet);
```
Comment (akshaymohite) [inline DALOptimizer/MatchExpr.cs]: Saw similar code in `VarDeclarationStmt` method as well. Can we refactor to avoid duplication?

**Review by akshaymohite — CHANGES_REQUESTED** [review]
@sushantyelpale suggested few changes. Please take care of them.

**Pull request merged**